tensorflow/lite/experimental/shlo/ops/dot_general.cc

```diff
@@ -0,0 +1,638 @@
+  const StorageT* rhs_data = rhs.GetDataAs<storage_type>();
+  StorageT* output_data = output.GetDataAs<storage_type>();
+  const DimensionSize lhs_size = lhs.NumElements();
+  const DimensionSize rhs_size = rhs.NumElements();
+  const DimensionSize output_size = output.NumElements();
+  const size_t lhs_rank = lhs.Rank();
+  const size_t rhs_rank = rhs.Rank();
+  const size_t output_rank = output.Rank();
+  const DimensionSize lhsb_size = lhs_batching_dimensions.size();
+  const DimensionSize lhsc_size = lhs_contracting_dimensions.size();
+
+  // function to generate indices for output
+  auto GenerateIndices = [&](size_t index) -> void {
+    size_t rank = op.output_shape.size();
+    size_t divisor = 1;
+    for (size_t i = 0, j = rank - 1; i < rank; ++i, --j) {
+      op.output_index[j] = (index / divisor) % op.output_shape[j];
+      divisor *= op.output_shape[j];
+    }
+  };
+  // function to incremement lhs and rhs indices
+  auto IncrementIndices = [&]() -> bool {
+    if (lhsc_size == 0) return false;
+    for (DimensionSize i = lhsc_size - 1; i >= 0; --i) {
+      op.lhs_index[lhs_contracting_dimensions[i]]++;
+      op.rhs_index[rhs_contracting_dimensions[i]]++;
+      if (op.lhs_index[lhs_contracting_dimensions[i]] <
+          lhs.shape().Dim(lhs_contracting_dimensions[i]))
+        return true;
+      if (i == 0) return false;
+      op.lhs_index[lhs_contracting_dimensions[i]] = 0;
+      op.rhs_index[rhs_contracting_dimensions[i]] = 0;
+    }
+    return true;
+  };
+  // pre compute helper for lhs and rhs indices
+  DimensionSize lhs_dim_accumulator = 1, rhs_dim_accumulator = 1;
+  for (size_t i = 0; i < lhs_rank; ++i) {
+    lhs_dim_accumulator *= lhs.shape().Dim(i);
+    op.lhs_index_helper[i] = lhs_size / lhs_dim_accumulator;
+  }
+  for (size_t i = 0; i < rhs_rank; ++i) {
+    rhs_dim_accumulator *= rhs.shape().Dim(i);
+    op.rhs_index_helper[i] = rhs_size / rhs_dim_accumulator;
+  }
+
+  StorageT output_element(0);
+  DimensionSize lhs_element_index = 0, rhs_element_index = 0;
+  for (DimensionSize k = 0; k < output_size; ++k, ++output_data) {
+    GenerateIndices(k);
+    absl::c_fill(op.lhs_index, 0);
+    absl::c_fill(op.rhs_index, 0);
+    size_t result_dim = 0;
+    for (size_t i = 0; i < lhsb_size; ++i, ++result_dim) {
+      op.lhs_index[lhs_batching_dimensions[i]] = op.output_index[result_dim];
+      op.rhs_index[rhs_batching_dimensions[i]] = op.output_index[result_dim];
+    }
+    for (size_t i = 0; i < op.lhs_result_dims.size(); ++i, ++result_dim) {
+      op.lhs_index[op.lhs_result_dims[i]] = op.output_index[result_dim];
+    }
+    for (size_t i = 0; i < op.rhs_result_dims.size(); ++i, ++result_dim) {
+      op.rhs_index[op.rhs_result_dims[i]] = op.output_index[result_dim];
+    }
+    output_element = 0;
+    while (true) {
+      lhs_element_index = 0;
+      rhs_element_index = 0;
+      for (size_t i = 0; i < lhs_rank; ++i) {
+        lhs_element_index += op.lhs_index[i] * op.lhs_index_helper[i];
```

Comment:
We should add an accessor to Tensor for addressing an element by a full index. We just didn't have a need for that before with elementwise ops, but it will certainly be the case for more complicated ones like this.

Reply:
Added Tensor::Get and Tensor::Set library methods. now we can use the get and set function to access the data or set the data.